rustc: Implement custom derive (macros 1.1)

alexcrichton opened the pull request:
This commit is an implementation of [RFC 1681] which adds support to the
compiler for first-class user-define custom `#[derive]` modes with a far more
stable API than plugins have today.

[RFC 1681]: https://github.com/rust-lang/rfcs/blob/master/text/1681-macros-1.1.md

The main features added by this commit are:

* A new `rustc-macro` crate-type. This crate type represents one which will
  provide custom `derive` implementations and perhaps eventually flower into the
  implementation of macros 2.0 as well.

* A new `rustc_macro` crate in the standard distribution. This crate will
  provide the runtime interface between macro crates and the compiler. The API
  here is particularly conservative right now but has quite a bit of room to
  expand into any manner of APIs required by macro authors.

* The ability to load new derive modes through the `#[macro_use]` annotations on
  other crates.

All support added here is gated behind the `rustc_macro` feature gate, both for
the library support (the `rustc_macro` crate) as well as the language features.

There are a few minor differences from the implementation outlined in the RFC,
such as the `rustc_macro` crate being available as a dylib and all symbols are
`dlsym`'d directly instead of having a shim compiled. These should only affect
the implementation, however, not the public interface.

This commit also ended up touching a lot of code related to `#[derive]`, making
a few notable changes:

* Recognized derive attributes are no longer desugared to `derive_Foo`. Wasn't
  sure how to keep this behavior and *not* expose it to custom derive.

* Derive attributes no longer have access to unstable features by default, they
  have to opt in on a granular level.

* The `derive(Copy,Clone)` optimization is now done through another "obscure
  attribute" which is just intended to ferry along in the compiler that such an
  optimization is possible. The `derive(PartialEq,Eq)` optimization was also
  updated to do something similar.

---

One part of this PR which needs to be improved before stabilizing are the errors
and exact interfaces here. The error messages are relatively poor quality and
there are surprising spects of this such as `#[derive(PartialEq, Eq, MyTrait)]`
not working by default. The custom attributes added by the compiler end up
becoming unstable again when going through a custom impl.

Hopefully though this is enough to start allowing experimentation on crates.io!

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
r? @nrc 

cc @cgswords

lifthrasiir commented on src/librustc_driver/driver.rs:
```diff
@@ -947,6 +947,10 @@ pub fn phase_3_run_analysis_passes<'tcx, F, R>(sess: &'tcx Session,
             middle::dead::check_crate(tcx, &analysis.access_levels);
         });
 
+        time(time_passes, "rustc-maro checking", || {
```
Typo.

eddyb commented on src/librustc_macro/lib.rs:
```diff
@@ -0,0 +1,164 @@
+#![feature(rustc_private)]
+#![feature(staged_api)]
+#![feature(lang_items)]
+
```
Isn't this crate what `proc_macro` is supposed to be?

alexcrichton replied:
Perhaps? I believe @nrc would know more

nrc replied:
Yeah, libproc_macro already provides a bunch of this functionality, I'd prefer to rename that and make changes rather than start from scratch here.

alexcrichton replied:
I wonder if it's perhaps best to start out conservatively? If we start out with a large list of unstable APIs here there'll be a lot of pressure to stabilize them, but starting with a much smaller base may help alleviate some of that pressure.

Perhaps we could look to merging the two after the initial implementation lands?

eddyb replied:
There's not *supposed* to be 2. I'd rather have the larger API surface commented out, or imported onl in a private module, and change that later, than have two crates which are the same.

alexcrichton replied:
This is already quite a large PR, however, and I don't have a great idea of where these crates are going. Perhaps it'd be best to defer to happening after this?

cgswords replied:
This file definitely needs to go into libproc_macro. A few other things:

- TokenStreams are already defined in `libsyntax/tokenstream.rs`, and that data structure should be consistent with this one.
- `lex` in `libproc_macro/parse.rs`already supports parsing a string into a tokenstream, so the `FromStr` just needs to wrap that.

alexcrichton replied:
To be clear this file is *tiny*, there's practically nothing here. Just `fmt::Display` and `FromStr`, as per [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/1681-macros-1.1.md).

Yes a new `TokenStream` type was added here, but it's never intended to duplicate functionality with the `TokenStream` already in libsyntax. Just wrap it in an intentionally unstable API and be conservative about what's exposed.

Currently `#[derive]` does not work at all on token streams, so the implementation detail doesn't use token streams yet. It's intended, though, that once we alter that (and add an attribute-like macro which takes token streams) we can move over to internally containing a `TokenStream`.

To clarify, though, do we have any way in-tree to go from an AST to a token stream without round tripping through a string?

cgswords replied:
The implementation detail sorta matters in that this `TokenStream` is going to quickly be a public, stabilized interface. If the entire idea is to stop breaking things, stabilizing this representation as a `TokenStream` isn't exactly the right course (unless we actually intend to discount the other one).

To clarify, no, we don't.

bors commented:
:umbrella: The latest upstream changes (presumably #35764) made this pull request unmergeable. Please resolve the merge conflicts.

nrc commented:
So, I'm not sure what to do with this PR - there is a lot I'd like to change, and I'm not sure if that will be easier to do before or after landing (it is already quite large, as you say). Ideally we'd fix stuff up before landing, but I expect a lot of things won't affect the API and it will allow users to start iterating if we land early. I also don't want to hold up landing too much since it is going to be a pain to rebase and is important to move quickly here.

So, issues:

* duplication of libproc_macro/librustc_macro - I think that we are starting small with proc_macro and it is going in the direction we want for macros 2.0. We have put quite a lot of effort into the design there and it seems a shame to ignore that. If you think some parts of that are too much initially, we can hide some of it in various ways. I think there is nearly enough in proc_macro to handle the 1.1 plan (precisely, there is enough with a PR I have which adds `to_string` for `TokenStreams`).

* Implementation of TokenStream - should really be tokens rather than AST, that seems quite a big thing to deal with, although it is purely internal. This point is pretty much subsumed by the previous one though.

* The dlsym strategy - I think it is better to use the existing macro infrastructure here. In particular, because it means we don't need to any ad hoc type checking code - the signature of the macro function is checked against a trait. Long-term, we also want to use a trait so we can allow macro authors to supply extra code or context about their macro. I have a PR which adds a new kind of Macro based on TokenStreams - I'd hoped to post that today, but I have been having testing difficulties.

All of these are things we can fix later, but would be easier to fix on a branch. So, I see three options:

* land this PR as is and fix things up later
* land my PR, then rebase this one to use the new macro kind and the existing TokenStream stuff
* split up this PR and land parts that standalone (e.g., the new crate type, the derive work) now and the other bits later.

All else being equal I'd prefer the last option. However, if clients are eager to get stuck into this asap and/or it will be really painful, then I don't mind too much going for the first option.

I've had a read through the code and other than the points above, I'm pretty happy with it. If we do go for option 1, then I'd like to do a detailed pass, but there didn't seem to be much point in doing that just yet if we go for another option.

eddyb commented:
> I think it is better to use the existing macro infrastructure here.

I disagree, other than in the interest of saving time. The current plugin-oriented infrastructure makes a lot of assumptions about the simple nature of name resolution in macros 1.0, whereas using metadata is more akin to the behavior of existing item hierarchies.
Furthermore, I believe it's good to distance procedural macro crates from rustc plugins, and to prevent them from being loaded as plugins.

aturon commented:
@nrc @alexcrichton

OK, it's seeming like there's a lack of communication and breakdown of process here.

In particular:

> duplication of libproc_macro/librustc_macro - I think that we are starting small with proc_macro and it is going in the direction we want for macros 2.0. We have put quite a lot of effort into the design there and it seems a shame to ignore that.

At the moment, the only accepted RFC we have on this topic is [Macros 1.1](https://github.com/rust-lang/rfcs/pull/1681), which was very precise about the naming and APIs that would be included, and the stabilization path. In contrast, the macros 2.0 direction here is currently only in [blog post form](http://www.ncameron.org/blog/libmacro/), and a [PR establishing libproc_macro](https://github.com/rust-lang/rust/pull/35538/) with *no accepted RFC*.

I agree with @nrc that these two things need to be unified. But I am at a loss as to why this was never brought up in the comment thread for the Macros 1.1 RFC. AFAICT, the implementation in this PR matches precisely what the RFC said, which was very carefully written to ensure everyone was on the same page about the API details, including the name of the crate. Why weren't these issues raised at that point? I'm also a little surprised that we have started to land pieces of the larger libproc_macro without an RFC accepted about experimentation there (or even a check-in with the relevant teams). That's not normally how we go about adding major new unstable API surfaces.

So, that's where we are. The question now is, what's the best way out of this quagmire?

Personally, I think we should:

- Very quickly come to consensus on any changes that are needed on Macros 1.1 on this thread. It sounds like it may just be the name of the crate? (Which I thought was already bikeshedded quite a bit...)

- Examine the plausibility of merging the two crates. It seems like @nrc and @alexcrichton (and possibly @cswords) should set up a call ASAP to try to sort this out.

- Ideally, land this PR with that unification in place -- unless that requires a large amount of rework. (I don't know the details well enough to say).

aturon commented:
(I should say that conversely, I'm was also surprised that this PR ignores the existence of libproc_macro; but OTOH it seems that very few people even realized that libproc_macro had landed, myself included. Like I said, there seems to be a process/communication breakdown prior to this PR.)

cgswords commented:
For what it's worth, there was a lot of dissent about the name of the initial macro crate [[1](https://github.com/rust-lang/rfcs/pull/1681#issuecomment-233453412)] [[2](https://github.com/rust-lang/rfcs/pull/1681#issuecomment-239558657)] [[3](https://github.com/rust-lang/rfcs/pull/1681#issuecomment-239765566)]. @nrc and I setteld on `libproc_macro` as the least of all evils in a side discussion, but we definitely should have done more (eg talked with @alexcrichton / commented more on the PR).

As for the landing stuff, part of the 'lack-of-RFC' thing was that I spent a decent chunk of my internship on the project, and there was a push to get it landed the compiler before my internship was over (which happened just barely in time).

As for actual tokensterams, there has been a `tokenstream` implementation in `libsyntax` for a few months now, and regardless of the ultimate fate of `libproc_macro`/`librustc_macro`, Macros 1.1 should probably lean on that whenever possible.

aturon commented:
> As for the landing stuff, part of the 'lack-of-RFC' thing was that I spent a decent chunk of my internship on the project, and there was a push to get it landed the compiler before my internship was over (which happened just barely in time).

Yeah, I certainly understand that, and I don't think *you* are at fault :)

We do sometimes allow landing experimental things without an RFC, but we usually have *some* kind of heads-up to the relevant teams and consensus that the experiment is OK, as well as a broadcast to the larger community about what's happening. It's probably obvious why: once something lands, there's significant inertia around it, so if we really want to operate in a consensus-based way we need consensus even on experiments.

Anyway, the people overseeing/reviewing this work should have ensured that communication happened, and planned ahead for the inevitable last-minute rush.

It's water under the bridge at this point, regardless.

alexcrichton commented:
Yes as @aturon mentioned I was hoping to just follow the design in the RFC, especially the contentious points about the API. I'd personally prefer to not deviate that far from what was accepted.

I was also under the impression that we needed more plumbing to use something like libproc_macro for custom derive modes. To be clear, we're not talking about entirely new macro systems here. We're talking about one and only one feature, custom derive. The `#[derive]` attribute takes an AST today, and in lieu of entirely rewriting that as part of this PR, it was simply easier to stick with that representation for the internals.

Once we move everything to token streams then we can fuse the libraries and/or build one on the other, and once we figure out the *stable* naming I figure we can move everything to that. In the meantime, though, `rustc_macro` is the accepted name for this library, and I don't know off the top of my head how to use token streams internally (but would love to know how!).

cgswords commented:
To clarify my position (which, again, is opinion):

1. I appreciate the RFC process, but Macros 1.1 seemed to move much faster than most RFCs by *months*. (That's hypothesis, but I would guess data mining confirms it by a good margin.) Moreover, it isn't exactly like the name concern wasn't raised (see above comment); it simply wasn't addressed in the RFC process. And the suggestion that we temporarily stick with the `rustc_macro` name seems to invalidate the *whole drive for stabilization* -- if the goal here is permanent stability, then the final name is certainly up for grabs (cc @ubsan @nikomatsakis). Moreover, since it wasn't seriously discussed as part of the RFC process, it *must be discussed here*. To summarize: based on the 'fast and forward' stabilization strategy, once we land 1.1, that name sticks, and that's the ball game. And since it wasn't dealt with before, it *must* be dealt with here.

2. `TokenStream`s *already* exist in the compiler (one of which is 2+ months old). Extending the original with `impl`s here is the right solution; defining a second `TokenStream` in the compiler is confusing at best and catastrophic at worst for stable users. Publicly leaking an `ast::item` fixes the internals more than any Macros 2.0 proposal has (or likely ever will) suggest.

aturon commented:
@cgswords 

First, I'm sorry if I came down hard before. I'm mostly just frustrated about the lack of coordination for this PR, which is leading to friction and wasted work.

> I appreciate the RFC process, but Macros 1.1 seemed to move much faster than most RFCs by months. 

RFCs merge at a wide range of rates, depending on scope and priority. Narrow, high priority RFCs often move quickly; libs RFCs often move relatively quickly. I don't understand the relevance of this point, in any case.

> Moreover, it isn't exactly like the name concern wasn't raised (see above comment); it simply wasn't addressed in the RFC process.

Looking more closely at the history, you're correct: the name became an explicit "unresolved question", to be resolved prior to stabilization. However:

> Moreover, since it wasn't seriously discussed as part of the RFC process, it must be discussed here.

So, the other part of the process for landing features in Rust is stabilization, where we generally have a tracking issue where people can discuss final details before stabilization (and we have a widely-communicated final comment period, to ensure that everyone's aware). I believe the intent was to sort out the final naming there.

Even so, we do need an initial name, and I agree that `rustc_macro` is not an appropriate choice for what will eventually become our general procedural macros story. I propose we go with `proc_macro` for now, but immediately start up more in-depth discussion on the tracking issue. How does that sound?

> TokenStreams already exist in the compiler (at 2+ months old). Extending the original with impls here is the right solution; defining a second TokenStream in the compiler is confusing at best and catastrophic at worst for stable users. Publicly leaking an ast::item concretizes the internals more than any Macros 2.0 proposal has (or likely ever will) suggest.

I agree that the two should be sharing a definition in the long run. But nothing is leaking an `ast::item` here, AFAICT.

I think the question right now is an engineering choice: given that we'd like to land Macros 1.1 relatively quickly, what's the best option for internal representation *for now*? Here, I don't know enough of the details to really weigh in; if the existing `TokenStream` is close to having full support for Macros 1.1, then it does indeed seem correct to just finish that off.

@alexcrichton, can you please spell out in a bit more detail the rationale for:

1. Not implementing directly in terms of the existing `TokenStream` type? (It sounds like @nrc is close to having the missing ingredients? I wasn't able to find the PR he mentioned.)

2. Using a newtype wrapper in the case that this is just `TokenStream` from `libsyntax`? Can we just re-export?

====

The above touches on two of the three issues @nrc raised. The dlsym issue, I can't really speak to.

eddyb commented:
> Can we just re-export?

One thing to remember is that impls of stable traits on stable types are always stable (AFAIK), so any trait impl on `TokenStream` would have to be audited and considered a guarantee once stabilized.

aturon commented:
> One thing to remember is that impls of stable traits on stable types are always stable (AFAIK), so any trait impl on TokenStream would have to be audited and considered a guarantee once stabilized.

That's a good point. I'm not sure how much of a risk that will be for this particular case, but it does seem prudent to newtype and otherwise be very careful about the promises we're making here.

I would note that there's impetus to move *somewhat* quickly here, just in the sense that at a minimum this feature is 12 weeks out from hitting stable, and the earlier we can land it in nightly, the earlier we can start getting feedback.

aturon commented:
Gah, I somehow missed [@alexcrichton's comment](https://github.com/rust-lang/rust/pull/35957#issuecomment-242814285) which already spells out some of the impetus for working with ASTs today. So cancel that question :-)

alexcrichton commented:
Whoa yes ok I think this may need to slow down a bit. I'd like to clarify some things first:

@cgswords 

> And the suggestion that we temporarily stick with the `rustc_macro` name seems to invalidate the *whole drive for stabilization* 

I implement the name `rustc_macro` because **it's what was accepted in the RFC**. There is *nothing* new that is stable as part of this PR, and that is very intentionally so. As @aturon mentioned as this is an open question that will likely change when we actually stabilize.

The name has already been debated on the RFC. I was hoping to **not reopen** debate about the name on this PR (as it just blocks everyone using the feature). The library can be named whatever, but to be clear the name `rustc_macro` is not stable, nor is it intended to be stable. It's representing the current consensus on the RFC to ensure that this moves along quickly without getting held up over what's already been discussed.

> To summarize: based on the 'fast and forward' stabilization strategy, once we land 1.1, that name sticks, and that's the ball game.

To reiterate from above again, **nothing new is stable in this PR**. The discussion on the RFC likely means the name will be changed, and it will be changed before stabilizing. We can do this in a manner that doesn't break any existing `rustc_macro` users and we'll deal with that down the road.

> `TokenStream`s *already* exist in the compiler (at 2+ months old). 

Again, to be clear, I have absolutely no intention of supplanting this. It is a technical fact that **token streams are now how `#[derive]` works today**, and as a result the more natural representation *currently* is just an AST item. As clarified by @eddyb *this is not exposed in any way, shape, or form*.

I would want to move over to the internal `TokenStream` type as quickly as possible once we actually can. Currently I'm under the impression that it requires more refactoring in the macro system, and there's no need to wait for any of that refactoring for this feature to land.

In other words, we could land this today and then change basically everything about how `#[derive]` is expanded internally without breaking everyone, as was exactly the intention.

> Publicly leaking an `ast::item` concretizes the internals more than any Macros 2.0 proposal has (or likely ever will) suggest.

To reiterate again, this is not happening, nor is it ever planned to happen.

---

@aturon 

> Not implementing directly in terms of the existing TokenStream type?

Sure! Today `#[derive]` operates over ASTs, not token streams. I don't personally know of a way to convert an AST to a token stream. The second `#[derive]` works over token streams we can reexport libsyntax's `TokenStream` type if possible or just change the implementation of the stable API.

> Using a newtype wrapper in the case that this is just TokenStream from libsyntax?

This is somewhat covered above, primarily by the lack of conversion from an AST to a `TokenStream`. If this exists that can be done, but it still won't change the fact that the underlying interface is all ASTs, not token streams.

---

@nrc 

> The dlsym strategy - I think it is better to use the existing macro infrastructure here. In particular, because it means we don't need to any ad hoc type checking code.

I'm not sure I understand this? The `#[plugin_registrar]` function today isn't typechecked at all AFAIK.

cgswords commented:
@aturon @alexcrichton Er, I really didn't mean that comment to seem as harsh as it apparently came across. Sorry about that. :/ And also, I clearly didn't read the code closely enough. Sorry about that, too. Thanks for explaining why I was wrong, though; I'm definitely onboard with this PR now, and excited to have it land.

eddyb commented:
@cgswords Do you think the quasi-quoting (& eventually other tools) can be in their own plugin crate, with everything that's *not* a syntax extension in the regular `macro` crate?
Eventually the `macro` crate would cause its dependencies which are syntax extensions to load and export them, I would think.
Which gives another reason to go the non-`plugin_registrar` route: it would make such transitive dependencies actually doable, with eventual full module system interaction, because `rustc` can reason about individual expander functions it knows about from metadata, and which crates they come from.

nrc commented:
@eddyb 

> > I think it is better to use the existing macro infrastructure here.

> I disagree, other than in the interest of saving time. The current plugin-oriented infrastructure makes a lot of assumptions about the simple nature of name resolution in macros 1.0, whereas using metadata is more akin to the behavior of existing item hierarchies.
Furthermore, I believe it's good to distance procedural macro crates from rustc plugins, and to prevent them from being loaded as plugins.

Sorry, I was a bit vague in my statement and also misunderstood a bit what was going on here. I think where/how we store the compiled macro isn't too important and the approach here (dlsym/metadata) seems fine. What I don't like is checking the signature of the macro by adding custom code to type checking. We can do what we do for MultiModifier, etc. and use a trait implementation for a function and use regular type checking to enforce the signature. Having read the PR in a bit more detail, this bit seems more orthogonal to the storage of the macro and the implementation of derive than I first thought.

I think name resolution and the way we store the macros is mostly orthogonal, either way there seems to be some degree of work to do to store macros in either fashion for plugins 2.0.

eddyb commented:
@nrc You can't do much better other than by enforcing a trait obligation instead of checking the individual type parts of the signature, i.e. [something like this](https://github.com/rust-lang/rust/blob/e4791e086d671d429db864787f6b60547a28b0f5/src/librustc_mir/transform/qualify_consts.rs#L1015-L1026).

nrc commented:
@aturon 

> OK, it's seeming like there's a lack of communication and breakdown of process here.

Yes. Sorry for my part of this.

> At the moment, the only accepted RFC we have on this topic is Macros 1.1, which was very precise about the naming and APIs that would be included, and the stabilization path.

I certainly misunderstood parts of the RFC, it seems. The spec for librustc_macro's TokenStreams doesn't have any implementation, and given the reference to the existing TokenStream, I assumed the implementation would be simply wrapping that.

More generally, my reading of the RFC was from the perspective of a maximally minimal slice of the proposed procedural macros functionality, and I assumed the implementation would follow that philosophy.

> In contrast, the macros 2.0 direction here is currently only in blog post form, and a PR establishing libproc_macro with no accepted RFC.

We talked explicitly about this (when discussing my q2 goals, iirc) and you told me we should experiment first *before* writing a libmacro RFC and that the level of detail in the blog post was enough to start that experimentation. My suggestion was to write an RFC and you told me not to. Perhaps I misunderstood where this experimentation should take place or you've changed your mind since the macros 1.1 RFC, but I thought evolving a small libmacro in-tree is the only way to do this.

There is really only the beginning of the library there and I didn't think it was worth announcing until we had a TokenStream-based macro form to use it with. I missed an opportunity to announce at least to internals that *something* had landed, but I didn't know there was any other macro work going on, so didn't think it was urgent.

> I agree with @nrc that these two things need to be unified. But I am at a loss as to why this was never brought up in the comment thread for the Macros 1.1 RFC. AFAICT, the implementation in this PR matches precisely what the RFC said, which was very carefully written to ensure everyone was on the same page about the API details, including the name of the crate. Why weren't these issues raised at that point?

I don't think there is much difference of opinion on the API here, only in the implementation, which is not really touched on in the RFC beyond some discussion of the crate types and how macros are stored. In particular the implementation for TokenStreams is not given and I assumed we'd discuss this after the RFC was accepted and before implementation (which is usually what happens with RFCs, I think it is very rare to bring up implmentation details in an RFC thread).

From the RFC:

> The actual underlying representation of `TokenStream` will be basically the same
> as it is in the compiler today. (the details on this are a little light
> intentionally, shouldn't be much need to go into *too* much detail).

> I'm also a little surprised that we have started to land pieces of the larger libproc_macro without an RFC accepted about experimentation there (or even a check-in with the relevant teams). That's not normally how we go about adding major new unstable API surfaces.

See above.

> So, that's where we are. The question now is, what's the best way out of this quagmire?

> *  Very quickly come to consensus on any changes that are needed on Macros 1.1 on this thread. It sounds like it may just be the name of the crate? (Which I thought was already bikeshedded quite a bit...)

I'm not actually too concerned about the name itself, more that we don't have two separate crates (my first proposal in this thread was to change the name of the existing libproc_macro to librustc_macro and merge in what is in this PR). We could also keep two separate crates, but have rustc_macro re-export a subset of proc_macro (since I think the two crates have different long term goals).

> *  Examine the plausibility of merging the two crates. It seems like @nrc and @alexcrichton (and possibly @cswords) should set up a call ASAP to try to sort this out.

This is important, and we should.

> * Ideally, land this PR with that unification in place -- unless that requires a large amount of rework. (I don't know the details well enough to say).

Agreed. But I'm also happy to land as is, as long as there is momentum to address the issue quickly, rather than leave it open once we have something that clients like Serde can use.

nrc commented:
@alexcrichton 

> I was also under the impression that we needed more plumbing to use something like libproc_macro for custom derive modes. To be clear, we're not talking about entirely new macro systems here. We're talking about one and only one feature, custom derive. The #[derive] attribute takes an AST today, and in lieu of entirely rewriting that as part of this PR, it was simply easier to stick with that representation for the internals.

Hmm, so this seems to be key to the choice of implementation of TokenStream. We certainly have the constraints that the clients must manipulate Strings and the deriving infrastructure wants to see AST. I agree we should avoid changing too much of that. However, it seems that this does not necessarily inform the implementation of TokenStream - we must go from AST to String to AST somewhere, but I don't think it is necessarily easier to do it one way vs another. In particular in this PR we go from String to AST in `from_str`, whereas I imagined that we would go from string to tokens in `from_str` and from tokens to AST in the custom derive glue (src/libsyntax_ext/deriving/custom.rs).

nrc commented:
@alexcrichton 

>@nrc

>>    The dlsym strategy - I think it is better to use the existing macro infrastructure here. In particular, because it means we don't need to any ad hoc type checking code.

> I'm not sure I understand this? The #[plugin_registrar] function today isn't typechecked at all AFAIK.

When a macro is registered with the registrar, its implementation (usually a function) must satisfy a trait bound, that only happens if the function signature matches the function type for which there is an `impl` of the syntax extension trait. I imagined that we could do something similar here, except that the register call would not be needed, it would effectively be desugared by the compiler.

nrc commented:
@eddyb 

> @nrc You can't do much better other than by enforcing a trait obligation instead of checking the individual type parts of the signature, i.e. something like this.

But I don't want any checking like this. I'm imagining something like:

```
fn registrar() {
    register(my_macro);
}

fn my_macro(TokenStream) -> TokenStream { ... }
```

where `fn <T: ProcMacro>register(macro: T) { ... }` and there is an `impl` of `ProcMacro` for functions with `TokenStream -> TokenStream` signature. Then, we treat the macro attribute on `my_macro` as a desugaring to the above, and the type checking is taken care of by the bounds check that `my_macro` implements `ProcMacro`.

I believe this is equivalent to the code you linked, however, here we're doing it syntactically without changing the type checker at all.

eddyb commented:
Oh, you're suggesting there'd be syntactical code generation involved, got it.

aturon commented:
@nrc 

> you told me we should experiment first before writing a libmacro RFC and that the level of detail in the blog post was enough to start that experimentation

So let me say, again, sorry for coming down a bit harshly with my first comment. We don't have a lot of precedent for this kind of situation, and so it's not surprising that we're hitting coordination issues.

But I stand by my original thought here: I think experimentation for proc_macro is good, and trying to lay everything out in an RFC before writing the code is probably not the best route. The issue I was trying to get at was two-fold. 

First, it's important that the lang and compiler teams as a whole have OK'ed this experimentation happening, and that there's a fairly clear scope for what we'll try. I don't think this has happened, and it'd be good to do it and have that record somewhere. (Note: *I'm* OK with it and think it's the right direction, but I'm not the only stakeholder here.)

Second, more pertinent to this PR, I was trying to point out that if there's a conflict between an approved RFC and plans in a blog post, it seems like the approved RFC should take precedence or be revisited.

Anyway, it seems like we're on track to a plan with this PR, so on procedural issues I think the next steps are:

- Get a robust discussion on the naming going on the tracking issue, so that we can quickly move to it and be in a good position to stabilize.

- Make sure the lang and compiler teams are aware of and on board with the proc_macro experimentation, ideally with a public record of the planned steps here (i.e., what's in scope for the experiment, at what point we'd expect an RFC, etc.)

nrc commented:
@aturon 

> Note: I'm OK with it and think it's the right direction, but I'm not the only stakeholder here

OK, I think this is where my misunderstanding arose.

> I was trying to point out that if there's a conflict between an approved RFC and plans in a blog post, it seems like the approved RFC should take precedence or be revisited.

Agreed, although in this case, I don't think there is conflict so much as 'merge conflict' - neither plan specified how it would integrate with the other and we need to work that out.

> Get a robust discussion on the naming going on the tracking issue, so that we can quickly move to it and be in a good position to stabilize.

I think there is some misunderstanding here - I think librustc_macro is fine and we should stick with it, but we need to decide on its relationship with the existing libproc_macro - should the new absorb the old or wrap it or re-export bits of it? We should of course still have *that* discussion.

> Make sure the lang and compiler teams are aware of and on board with the proc_macro experimentation, ideally with a public record of the planned steps here (i.e., what's in scope for the experiment, at what point we'd expect an RFC, etc.)

I'll hash out an internals post this week.

nrc commented:
@aturon (et al)

> the API surface that we stabilize for 1.1 is not intended for deprecation

ah, ok, I misunderstood, I assumed this would be deprecated once procedural macros were properly stood up and stabilised. In which case, we should think about other names...

nrc commented:
btw, the branch/PR I was hinting at is almost done - I need to add tests and iron a wrinkle or two - hopefully it will be ready tomorrow. It adds a `ProcMacro` kind of SyntaxExtension to the existing system to allow 'old' plugin-style macros with a `TokenStream -> TokenStream` signature. I don't think it overlaps with anything in this PR, but should make it easier to integrate the two approaches.

dtolnay commented:
I have a Serde implementation at https://github.com/serde-rs/serde/pull/530. It is obnoxious to build without support from Cargo but it does work!

- The interesting part: [`serde_derive/src/lib.rs`](https://github.com/serde-rs/serde/blob/d914fdf67b175b95168f163727578ac0b8d6883b/serde_derive/src/lib.rs)
- How to try it, if you dare: [`tmp_test/build.sh`](https://github.com/serde-rs/serde/blob/d914fdf67b175b95168f163727578ac0b8d6883b/tmp_test/build.sh)

One bug to report so far: the `#![feature(rustc_macro)]` in the client crate (not in the rustc-macro crate) is necessary but is not getting marked as used.

```
warning: unused or unknown feature, #[warn(unused_features)] on by default
 --> src/main.rs:1:12
  |
1 | #![feature(rustc_macro)]
  |            ^^^^^^^^^^^
```

dtolnay commented:
Another bug: if I use macros (in my case `try!`) in my generated code, it fails after returning the expanded TokenStream.

```
thread 'rustc' panicked at 'fold_mac disabled by default', src/libsyntax/fold.rs:171
stack backtrace:
   1:     0x7f8249c8c843 - std::sys::backtrace::tracing::imp::write::hacf40ae6a3c9a7e3
   2:     0x7f8249c9e94c - std::panicking::default_hook::_{{closure}}::ha6b33fef3a323878
   3:     0x7f8249c9cbc7 - std::panicking::default_hook::h7c2212d31b24a51e
   4:     0x7f8249c9d3d8 - std::panicking::rust_panic_with_hook::h4841ecb3443c75a3
   5:     0x7f824738a91f - std::panicking::begin_panic::h9979c691b33ecb57
   6:     0x7f82473bf520 - syntax::fold::noop_fold_expr::habcd8d7dbabeb7d9
   7:     0x7f8247387c60 - _<syntax..ptr..P<T>>::map::h6b45552e12df359d
   8:     0x7f82473b7351 - syntax::fold::noop_fold_expr::habcd8d7dbabeb7d9
   9:     0x7f8247387c60 - _<syntax..ptr..P<T>>::map::h6b45552e12df359d
  10:     0x7f82473c58e4 - syntax::fold::noop_fold_stmt::hc2a59bca18569e1c
  11:     0x7f82473de320 - _<collections..vec..Vec<T> as syntax..util..move_map..MoveMap<T>>::move_flat_map::h816fb3c8065e291d
  12:     0x7f8247387b3e - _<syntax..ptr..P<T>>::map::h376898895ca90444
  13:     0x7f82473c73df - syntax::fold::noop_fold_impl_item::h2c1153c2a25211ef
  14:     0x7f82473c8296 - syntax::fold::noop_fold_item_kind::h152671aac09b61be
  15:     0x7f82473c43e5 - syntax::fold::noop_fold_item::h2f2f9663a2fafba2
  16:     0x7f82473c54a3 - syntax::fold::noop_fold_stmt::hc2a59bca18569e1c
  17:     0x7f82473de320 - _<collections..vec..Vec<T> as syntax..util..move_map..MoveMap<T>>::move_flat_map::h816fb3c8065e291d
  18:     0x7f8247387b3e - _<syntax..ptr..P<T>>::map::h376898895ca90444
  19:     0x7f82473c73df - syntax::fold::noop_fold_impl_item::h2c1153c2a25211ef
  20:     0x7f82473c8296 - syntax::fold::noop_fold_item_kind::h152671aac09b61be
  21:     0x7f82473c43e5 - syntax::fold::noop_fold_item::h2f2f9663a2fafba2
  22:     0x7f82473c54a3 - syntax::fold::noop_fold_stmt::hc2a59bca18569e1c
  23:     0x7f82473de320 - _<collections..vec..Vec<T> as syntax..util..move_map..MoveMap<T>>::move_flat_map::h816fb3c8065e291d
  24:     0x7f82473bb18c - syntax::fold::noop_fold_expr::habcd8d7dbabeb7d9
  25:     0x7f8247387c60 - _<syntax..ptr..P<T>>::map::h6b45552e12df359d
  26:     0x7f82473c9303 - syntax::fold::noop_fold_item_kind::h152671aac09b61be
  27:     0x7f82473c43e5 - syntax::fold::noop_fold_item::h2f2f9663a2fafba2
  28:     0x7f82473d7967 - _<core..iter..Map<I, F> as core..iter..iterator..Iterator>::next::h249d979a8cc8c186
  29:     0x7f82473f3fbf - _<syntax_ext..deriving..custom..CustomDerive as syntax..ext..base..MultiItemModifier>::expand::hf349ab3c8874c5dc
  30:     0x7f824740a2c2 - syntax_ext::deriving::expand_derive::he9febb5a624fc3de
  31:     0x7f824739fd99 - _<F as syntax..ext..base..MultiItemModifier>::expand::h38940f50d2a289b6
  32:     0x7f8243d1578f - syntax::ext::expand::expand_annotatable::h3f215f12041ec02d
  33:     0x7f8243d12dd8 - syntax::ext::expand::expand_item::haedddad4859a8cd1
  34:     0x7f8243d22f94 - _<syntax..ext..expand..MacroExpander<'a, 'b> as syntax..fold..Folder>::fold_item::h539d4f688f0a95be
  35:     0x7f8243bec4b3 - syntax::fold::noop_fold_mod::hb59fc673d8e45f3e
  36:     0x7f8243be2fbd - syntax::fold::noop_fold_item_kind::hd5ac5113f1eb0167
  37:     0x7f8243becd7a - syntax::fold::noop_fold_item_simple::hd942d2dbe39b4c73
  38:     0x7f8243d1757f - syntax::ext::expand::expand_annotatable::h3f215f12041ec02d
  39:     0x7f8243d12dd8 - syntax::ext::expand::expand_item::haedddad4859a8cd1
  40:     0x7f8243d23080 - _<syntax..ext..expand..MacroExpander<'a, 'b> as syntax..fold..Folder>::fold_item::h539d4f688f0a95be
  41:     0x7f8243d220b3 - _<syntax..ext..expand..MacroExpander<'a, 'b> as syntax..fold..Folder>::fold_crate::h5e77d41aa2154739
  42:     0x7f8243d245c7 - syntax::ext::expand::expand_crate_with_expander::hbf86abd258a1b039
  43:     0x7f8243d240fb - syntax::ext::expand::expand_crate::he4ea5351a24e7545
  44:     0x7f824a08f217 - rustc_driver::driver::phase_2_configure_and_expand::_{{closure}}::ha8d147f1b97c1861
  45:     0x7f824a03da0d - rustc_driver::driver::phase_2_configure_and_expand::hb1da6e0a28c34539
  46:     0x7f824a033fc2 - rustc_driver::driver::compile_input::h3963b192268fd661
  47:     0x7f824a069916 - rustc_driver::run_compiler::h36bd4bd1981ac50f
  48:     0x7f8249fa3ef2 - std::panicking::try::do_call::hd90d510348f122a6
  49:     0x7f8249cae2f6 - __rust_maybe_catch_panic
  50:     0x7f8249fbdbca - _<F as alloc..boxed..FnBox<A>>::call_box::h3e8653068bf4b5ed
  51:     0x7f8249c9af92 - std::sys::thread::Thread::new::thread_start::hcaa3a549050fafca
  52:     0x7f82430556f9 - start_thread
  53:     0x7f82498dab5c - clone
  54:                0x0 - <unknown>
```

alexcrichton commented:
Ok I'm kinda lost at this point in terms of what needs to be done. I just pushed a new implementation which uses syntactical generation (aka what libtest does) rather than verification of the types manually, but other than that all the infrastructure is still the same.

@nrc were there more pieces you'd like changed? I can't really follow the discussion any more...

@dtolnay thanks for testing! I think I've fixed both those issues (and added tests)

nrc commented:
So, I think we could merge sooner rather than later. The outstanding issues are:

* the name of the librustc_macro crate (I attempted to start discussion over on the [tracking issue](https://github.com/rust-lang/rust/issues/35900#issuecomment-243253702),
* how to merge the existing libproc_macro and the new librustc_macro crates,
* whether to/how to integrate this work more closely with the existing macro system.

I'm happy to merge this PR first and then decide these issues, but I think we have to decide at least the naming issue before advertising the feature. Merging now and adding these to the tracking issue is probably the most efficient way to move here.

@alexcrichton, @aturon does that sound like a plan? If so I'll review the PR with a view to merging soon.

aturon commented:
@nrc I'm fine with merging now; hopefully we can address the naming issue quickly. (Though if we think libproc_macro is what we want, that'll force the issue of merging as well.)

alexcrichton commented:
> the name of the librustc_macro crate

I was personally hoping we can defer this debate to happen asynchronously wrt landing this. 

Also note that whatever name we land on we can have a graceful deprecation period for the name `rustc_macro`. In other words we can start messaging immediately and then not break everyone when the name changes.

> how to merge the existing libproc_macro and the new librustc_macro crates

Currently this actually isn't possible without some larger restructuring. proc_macro depends on rustc_plugin which depends on rustc_metadata. The rustc_metadata crate, however, now depends on rustc_macro so it knows what to do when it's loading types. Without some larger restructuring I'm not sure this would be easily possible? Unless you know how we might rejigger this in the meantime?

> whether to/how to integrate this work more closely with the existing macro system.

I initially tried to literally use `#[plugin_registrar]`, but the code paths for discovery of plugin crates vs discovered of crates with `#[macro_use]` are entirely different. It ended up looking like it would be a pretty big refactor to enable something like that where everything is funneled through one interface.

In other words, I'd be fine doing so, but I'd also prefer to rewrite reams of interfaces as part of this as well...

---

I guess all that's basically to say that I'd be in favor of landing now and working on the architectural issues in parallel. It looks like there's going to be a lot of refactoring needed to truly integrate this well, unfortunately, but that's also just another way of saying that macros 2.0 will take a lot of work!

nrc commented on src/librustc_trans/back/linker.rs:
```diff
@@ -456,7 +457,8 @@ fn exported_symbols(scx: &SharedCrateContext,
     // why we don't ever need dylib symbols on non-MSVC.
-    if crate_type == CrateType::CrateTypeDylib {
+    if crate_type == CrateType::CrateTypeDylib ||
+       crate_type == CrateType::CrateTypeRustcMacro {
```
could factor out a CrateType::dynamically_linked function

alexcrichton replied:
Ah this is actually kinda subtle b/c cydlib is also dynamically linked. I do agree though that the number of places that had to be updated with a new crate type was unfotunate :(

nrc commented on src/libsyntax/ext/expand.rs:
```diff
@@ -440,8 +440,7 @@ fn expand_annotatable(mut item: Annotatable, fld: &mut MacroExpander) -> SmallVe
                     span: Some(attr.span),
-                    // attributes can do whatever they like, for now
-                    allow_internal_unstable: true,
+                    allow_internal_unstable: false,
```
would this would be a breaking change?

alexcrichton replied:
I believe so, yes, but only a plugin-breaking-change. That is, all existing plugins today could silently rely on unstable features, but now they won't be able to unless they manufacture one of these spans.

Note, though, that if we *don't* change this then all `rustc-macro` derive modes would also be able to rely on unstable features :(

dtolnay commented:
@alexcrichton I confirmed that the two issues I reported are fixed, but now I need *both* of `#![feature(rustc_macro, rustc_macro_lib)]` in the crate containing rustc_macro_derive which I didn't before. Is that intentional? Can we pick one or the other?

eddyb commented:
@dtolnay Isn't one of those from the `extern crate rustc_macro;` and the other one from the crate-type?
I guess they could use the same feature-gate, not sure though.

alexcrichton commented:
@dtolnay unfortunately the compiler has a restriction right now that library feature names and language feature names do not intersect. I tried to hack around this before and give them the same feature name, but that was the bug about the spurious "unused feature" warning. In the end I figured that writing two features isn't so bad, and fixing that assumption in the compiler might be more effort than it's worth.

dtolnay commented:
Thanks for the explanation @alexcrichton. I haven't looked at the code but it seems to work based on my testing. The next thing I would look for is Cargo support and then I'll publish Serde's proc macro crate so we can test more broadly.

alexcrichton commented:
Definitely! I think the Cargo support here should be trivial, and I'll push on that as soon as this lands.

nrc commented:
r=me with a `[breaking-change]` note added to the commit message (I think though that the PR is large enough that it is not worth batching like we would with most libsyntax breaking changes).

alexcrichton commented:
@bors: r=nrc 

Thanks for the review @nrc!

bors commented:
:pushpin: Commit 2ab40cf has been approved by `nrc`

<!-- @bors r=nrc 2ab40cf4bba9f418e0ad8d4701e63dec5c59d246 -->

bors commented:
:umbrella: The latest upstream changes (presumably #34494) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@bors: r=nrc

dtolnay commented:
I filed #36211 as a followup which I expect will be easier to fix in rustc than for us to work around in Syntex.

bors commented:
:umbrella: The latest upstream changes (presumably #35894) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@bors: r=nrc

bors commented:
:pushpin: Commit e7cb646 has been approved by `nrc`

<!-- @bors r=nrc e7cb64617c43cdca0193eaf23e5b9c575ba68c34 -->

bors commented:
:hourglass: Testing commit e7cb646 with merge 3b5bca7...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2366)

dtolnay commented on src/librustc_macro/lib.rs:
```diff
@@ -0,0 +1,169 @@
+//! new custom derive modes through `#[rustc_macro_derive]`.
+//!
+//! Added recently as part of [RFC 1681] this crate is currently *unstable* and
+//! requires the `#![feature(rustc_macro)]` directive to use. Eventually,
```
This should be `rustc_macro_lib`.

alexcrichton commented:
@bors: r=nrc

bors commented:
:pushpin: Commit 66ec18f has been approved by `nrc`

<!-- @bors r=nrc 66ec18f6cbb72abafc73b04c08904b02cbce61ac -->

alexcrichton commented:
@bors: r- oh right haven't fixed tests yet.

alexcrichton commented:
@bors: r=nrc

bors commented:
:pushpin: Commit ecc6c39 has been approved by `nrc`

<!-- @bors r=nrc ecc6c39e876b69496bc88ef47ff3a339662346b1 -->

bors commented:
:hourglass: Testing commit ecc6c39 with merge a029ea3...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10411), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10465), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1699), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1524), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3683), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10405), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10366), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2338), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10343), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3565), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5515), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10444), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10417), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2362), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1541), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5531), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2393), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5507), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4700), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1718), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5512), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2388)
Approved by: nrc
Pushing a029ea343fe5c9b7cf550902b5a507cd8c22a833 to master...

kumbayo commented on src/libsyntax/feature_gate.rs:
```diff
@@ -302,7 +305,6 @@ declare_features! (
     (removed, struct_inherit, "1.0.0", None),
     (removed, test_removed_feature, "1.0.0", None),
     (removed, visible_private_types, "1.0.0", None),
-    (removed, unsafe_no_drop_flag, "1.0.0", None)
```
I am curious, was this line removed intentionally?

petrochenkov replied:
Heh, couple of pages above there's a comment `// Don't ever remove anything from this list; set them to 'Removed'.`, but people do this anyway.

alexcrichton replied:
Oops yes this was not intentional, just a botched merge.

eddyb replied:
FWIW library features do get removed all the time so I don't think we need to bother anymore with this list.
